### ISSUE TITLE
Masternode Account Security

### ISSUE THREAD

**Issue by CalRon07** [CPChain/chain#95]
Hi there,

This is CPChain team. Note that is page is for users willing to request new features. For bugs, please select *Bug Report* template.

**Is your feature request related to the chain, the wallet, the console, the explorer, or other aspect?**

The Wallet

**Is your feature request related to a problem? Please describe.**

I recommend CPChain to integrate a feature which allows RNodes to store their coins in a separate account not associated with their masternode in order to protect their coins.  

This feature would still give masternodes the same points as they would if they stored their coins in the account used for their RNode.

I have run masternodes for other projects, and I have been a victim of a stolen keystore which eventually led to stolen coins.  

This feature would help protect masternodes from having their coins stolen in the event their server was hacked.

**Describe the solution you'd like**

I run a masternode for High Performance Blockchain (HPB), and they came up with a solution to protect their masternodes from having their coins stolen.  

They developed a feature which allows masternodes to store their coins in a 'modified wallet'.  This wallet is linked to the masternode's account associated with their server (the binding address).  The coins in the modified wallet count as points towards a high performance node, just as the coins would if they were in the binding address.

This feature protects the masternode in the event their server was hacked and keystore stolen.

The modified wallet does not have the keystore or private key stored anywhere on the server, so if the server was hacked, the coins in the modified wallet would be safe.

The rewards for proposing blocks would be sent to the binding address (address associated with the server).

The reason why I recommend this is because I have been a victim of stolen coins due to the keystore for one of my masternodes being stolen.

**Describe alternatives you've considered**
None at this moment.

**Additional context**

HPB Masternodes asked HPB to come up with a feature to allow us to store our coins in a separate wallet not associated with our masternode, yet still give us points towards becoming a high performance node.  HPB was able to develop this feature and it has been virtually flawless.  

They would be a good point of contact in order to help CPChain develop this feature.

**Comment by hvredevoort** [comment]
That's a very sensible solution and I hope @alex031029 can implement this with the team
I already submitted several recommendations and feature improvements related to security but up to now the team has not really committed to anything. Security seems to be ultra low priority for CPChain, which is very odd because a lot of money is at stake.

**Comment by alex031029** [comment]
Hi there!

Firstly, I am sorry for your loss of tokens in other project. 

And from my perspective, this request is very reasonable, and I have discussed your request with the team as well as our CTO. Their all concurred that it could be a good feature to be embedded into the chain in the future. 

One major concern from the developer team is that this new feature could take quite a long time to implement. At least the following things have to be done:

1. The mechanism to reward. The current implementation to reward a node is to directly modify the balance, in which no transactions are involved. It mechanism is inherited from ETH, and different from BTC in which tokens are rewarded by transactions. 

2. Test cases. This new feature requires a very large amount of new test cases. 

In conclusion, we are seriously considering add this feature. But it won't be implemented very soon.

**Comment by CalRon07** [comment]
Thank you @alex031029 for the very fast response!  I am extremely happy to hear the team likes this idea.  I assure you this feature will save a lot of stress and heartache, and create a full team of secure RNodes.  

I understand developing a feature like this is time consuming, especially when your hands are already full with R&D.

Regarding concern #1 from the developer team:  
As far as I'm aware with HPB, when we receive rewards for mining blocks, there is no transaction involved.  Our account balance increases, but I have never seen a transaction for the rewards sent to my account.  If this is correct, maybe one your developers can contact HPB to see what process they used (I can provide a point of contact if you'd like)?  Hopefully this will help save some time for your developers.

Simply knowing CPC is interested in this feature is good enough for me.  Take as much time as needed.  I am here to help in any way I can.

Thank you very much!
